### PR TITLE
use router.ResolvedURL in more places

### PR DESCRIPTION
**Pull request by rogpeppe**
In particular, we make AddCharm and AddBundle and friends take a ResolvedURL.
Almost all of the changes are in the tests.

### CODE REVIEW

File: csclient/csclient_test.go [hunk 1]
```diff
@@ -1051,7 +1042,7 @@ func (s *suite) TestMacaroonAuthorization(c *gc.C) {
 	body, err := json.Marshal([]string{"bob"})
 	c.Assert(err, gc.IsNil)
 
-	resp, err := s.client.DoWithBody(req, "/"+purl.Path()+"/meta/perm/read", httpbakery.SeekerBody(bytes.NewReader(body)))
+	resp, err := s.client.DoWithBody(req, "/"+rurl.URL.Path()+"/meta/perm/read", httpbakery.SeekerBody(bytes.NewReader(body)))
```
Comment (mhilton) [inline csclient/csclient_test.go:1045]: This is a different thing purl == rurl.PreferredURL() In this case I suspect it doesn't matter.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/828/
Test PASSed.

File: internal/charmstore/store_test.go [hunk 2]
```diff
@@ -1006,6 +941,35 @@ func urlStrings(urls []*charm.Reference) []string {
 	return urlStrs
 }
 
+// mustParseResolvedURL parses a resolved URL in string form, with
+// the optional promulgated revision preceding the entity URL
+// separated by a space.
+func mustParseResolvedURL(urlStr string) *router.ResolvedURL {
+	s := strings.Split(urlStr, " ")
```
Comment (mhilton) [inline internal/charmstore/store_test.go]: strings.Fields?

File: internal/charmstore/store_test.go [hunk 3]
```diff
@@ -1006,6 +941,35 @@ func urlStrings(urls []*charm.Reference) []string {
...
+	promRev := -1
+	if len(s) == 2 {
```
Comment (mhilton) [inline internal/charmstore/store_test.go]: switch len(s)?

File: internal/v4/archive.go [hunk 4]
```diff
@@ -517,17 +523,16 @@ func setArchiveCacheControl(h http.Header, idFullySpecified bool) {
 	h.Set("Cache-Control", "public, max-age="+strconv.Itoa(seconds))
 }
 
-// getPromulgated URL finds the promulgatedURL that should be used for
-// this newly uploaded charm, if the charm should be promulgated,
-// othewise it returns nil. An error is returned if there is a problem
-// communicating with the storage.
-func (h *Handler) getPromulgatedURL(id *charm.Reference) (*charm.Reference, error) {
+// getNewPromulgatedRevision URL returns the promulgated revision
```
Comment (mhilton) [inline internal/v4/archive.go]: s/ URL //

File: internal/v4/relations_test.go [hunk 5]
```diff
@@ -855,3 +833,24 @@ func (s metaAnyResponseById) Swap(i, j int) { s[i], s[j] = s[j], s[i] }
 func (s metaAnyResponseById) Less(i, j int) bool {
 	return s[i].Id.String() < s[j].Id.String()
 }
+
+// mustParseResolvedURL parses a resolved URL in string form, with
+// the optional promulgated revision preceding the entity URL
+// separated by a space.
+func mustParseResolvedURL(urlStr string) *router.ResolvedURL {
```
Comment (mhilton) [inline internal/v4/relations_test.go:840]: Is there a way to avoid having this function twice?

**Comment by mhilton** [comment]
:+1: LGTM, with a couple of things to consider

File: internal/v4/relations_test.go (hunk 5 above)
Comment (urosj) [inline internal/v4/relations_test.go:840]: +1

**Comment by urosj** [comment]
:+1:

File: internal/v4/archive.go (hunk 4 above)
Comment (rogpeppe) [inline internal/v4/archive.go]: done

File: internal/v4/relations_test.go (hunk 5 above)
Comment (rogpeppe) [inline internal/v4/relations_test.go:840]: probably. i will consider possibilities for a future PR. perhaps we could bless the GoString form of ResolvedURL and put the parsing code into router itself.

File: internal/charmstore/store_test.go (hunk 3 above)
Comment (rogpeppe) [inline internal/charmstore/store_test.go]: I considered it - it didn't seem that much better, so I've left it.

File: internal/charmstore/store_test.go (hunk 2 above)
Comment (rogpeppe) [inline internal/charmstore/store_test.go]: done. that makes it better to use a switch, so I've done that too.

File: csclient/csclient_test.go (hunk 1 above)
Comment (rogpeppe) [inline csclient/csclient_test.go:1045]: That's what I thought at the time.

**Comment by rogpeppe** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore/829/
Test PASSed.

**Pull request merged**